src/ng/directive/ngModel.js

```diff
@@ -810,7 +810,10 @@ var NgModelController = ['$scope', '$exceptionHandler', '$attrs', '$element', '$
 
     // if scope model value and ngModel value are out of sync
     // TODO(perf): why not move this to the action fn?
-    if (modelValue !== ctrl.$modelValue) {
+    // Second check is needed to allow setting the model to NaN when there's an asyncValidator
+    if (modelValue !== ctrl.$modelValue &&
+       (ctrl.$modelValue === ctrl.$modelValue || modelValue === modelValue)
```

Comment:
Quick tests shows that `isNaN` does not work, but `Number,isNaN` (ECMA6) does. :(